lib/trento/application/integration/discovery/payloads/cloud_discovery_payload.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Trento.Integration.Discovery.CloudDiscoveryPayload do
+  @moduledoc """
+  Cloud discovery integration event payload
+  """
+
+  @required_fields []
+
+  use Trento.Type
+  import PolymorphicEmbed, only: [cast_polymorphic_embed: 3]
+
+  deftype do
+    field :provider, Ecto.Enum, values: [:aws, :gcp, :azure, :unknown], default: :unknown
+
+    field :metadata, PolymorphicEmbed,
+      types: [azure: __MODULE__.AzureMetadata],
+      on_type_not_found: :nilify,
+      on_replace: :update
+  end
+
+  def changeset(event, attrs) do
+    enriched_attrs = enrich_metadata_type(attrs)
+
+    event
+    |> cast(enriched_attrs, [:provider])
+    |> cast_polymorphic_embed(:metadata, required: false)
```

Comment:
Ok, now I understand why `ExMachina` is not an issue here, as we use the polymorphism just as a decoder, and not in the read models. So in fact, we don't create factories from it

Reply:
Exactly! Fortunately here we can use polymorphic embed :smiling_face_with_tear: